### DISCUSSION TITLE
Integration your library to VSCode & ESP-IDF

### DISCUSSION THREAD

**Discussion by KiShiVi** [pschatzmann/ESP32-A2DP#563] [Q&A] [answered]
Hello! Thank you very much for your library. I used it regularly in Arudino IDE. Today I was faced with the task of transferring the project to VSCode & ESP-IDF. I did everything according to your instructions:
1) Made a clone for my project
2) Renamed the main file of the main project from "c" to "cpp" file
3) Fixed this in the CMake file.
4) Set up SDK configuration

When build the project, I get a bunch of errors related to the use of outdated FreeRTOS definitions, incorrect type casting in the ESP_LOGI function, and including the "algorithm" library.

The problem with FreeRTOS definitions was resolved by setting the "configENABLE_BACKWARD_COMPATIBILITY" parameter in the SDK configuration. I don’t know how to deal with type casts and other errors. Of course, I can do all the type conversions manually, but I don’t want to do this every time you update your library. Please tell me what I'm doing wrong.. Thank you!

main.cpp:
```
#include "BluetoothA2DPSink.h"
BluetoothA2DPSink a2dp_sink;

void setup(){
  a2dp_sink.start("MyMusic");  
}

void loop(){
   delay( 500 );
}


extern "C" void app_main(void){
  setup();
  while(true){
    loop();
  }
}
```

Errors:
```
D:/Espressif/v5.2.1/esp-idf/components/log/include/esp_log.h:291:27: error: format '%x' expects argument of type 'unsigned int', but argument 7 has type 'uint32_t' {aka 'long unsigned int'} [-Werror=format=]
  291 | #define LOG_COLOR(COLOR)  "\033[0;" COLOR "m"
D:/Espressif/v5.2.1/esp-idf/components/log/include/esp_log.h:291:27: error: format '%x' expects argument of type 'unsigned int', but argument 6 has type 'uint32_t' {aka 'long unsigned int'} [-Werror=format=]
  291 | #define LOG_COLOR(COLOR)  "\033[0;" COLOR "m"
D:/Git/bluetoothaudiospeaker/BluetoothAudioSpeaker/components/ESP32-A2DP/src/BluetoothA2DPSink.cpp:1365:1: error: format '%x' expects argument of type 'unsigned int', but argument 6 has type 'uint32_t' {aka 'long unsigned int'} [-Werror=format=]
 1365 |                rc->rmt_feats.feat_mask, rc->rmt_feats.ct_feat_flag);
      |                              ~
      |                              |
      |                              uint32_t {aka long unsigned int}
      
```
      

![image](https://github.com/pschatzmann/ESP32-A2DP/assets/50634934/b12037cf-a477-4e70-a8ba-aa0af394cefe)
![image](https://github.com/pschatzmann/ESP32-A2DP/assets/50634934/ed0004cc-614d-49cc-a333-364ad7112333)

**Comment by pschatzmann** [comment, marked as answer]
I committed some corrections. 

But please note that you need to decide which output functionality you want to use with the help of
A2DP_LEGACY_I2S_SUPPORT and A2DP_I2S_AUDIOTOOLS

**Reply by KiShiVi** [reply]
I saw your edits!  Thank you very much, I will definitely try to assemble the project again during the day!  Thank you!

**Reply by KiShiVi** [reply]
Good afternoon, thanks a lot for your edits - I managed to build a project for pure ESP-IDF in VSCode. In the process of debugging I faced the following problem - you may be interested in this:
When using ESP-IDF > 5.2.1, esp_bluedroid_config_t is not initialised with the default value (SSP = EN).

BluetoothA2DPCommon.cpp
```
esp_err_t BluetoothA2DPCommon::bluedroid_init(){
#if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(5, 2 , 1)
    return esp_bluedroid_init_with_cfg(&bluedroid_config);
#else
    return esp_bluedroid_init();
#endif
}
```

This problem leads to the following error:

```
E (1265) BT_GAP: esp_bt_gap_set_security_param is not supported when `ssp_en` in `esp_bluedroid_config_t` is disabled!
```

Although bluetooth starts and I can find ESP32 when scanning - it asks for a PIN, which is also not a standard one (0000 and 1234 do not fit).

This problem was solved by forcing the configuration before calling a2dp_sink.start(‘MyMusic’); using the method you recently implemented.

main.cpp
```
#include "BluetoothA2DPSink.h"
#include "freertos/task.h"

BluetoothA2DPSink a2dp_sink;

extern "C" void app_main(void){

  esp_bluedroid_config_t cfg {
    .ssp_en = true
  };
  a2dp_sink.set_bluedroid_config_t(cfg);
  
  a2dp_sink.start("MyMusic");  
  while(true){
    vTaskDelay(pdMS_TO_TICKS(100));
  }
}
```

**Reply by pschatzmann** [reply]
Thanks, I changed the default initialization to true
